lib/ecto/query/planner.ex

```diff
@@ -2027,14 +2027,29 @@ defmodule Ecto.Query.Planner do
   defp cte_fields([], [], _aliases), do: []
 
   defp assert_update!(%Ecto.Query{updates: updates} = query, operation) do
+    read_only_fields =
+      case get_source!(:updates, query, 0) do
+        {source, schema, _} when is_binary(source) and schema != nil ->
+          schema.__schema__(:read_only)
```

Comment:
Do we raise if the field is virtual? If we don't, should we? Maybe this should be called `non_writeable` instead and include both virtual and read only fields?

Reply:
It looks like it's handled when trying to get the type. The validation is shared with selects, etc.

But actually I like consolidating the two here and in other places. I will make the change it if that's alright with you

Reply:
@greg-rychlewski perhaps we could unify it by introducing `writeable_type`? This way we don't need to perform additional traversals.

Reply:
that's a really good idea. i will see what i can do

Reply:
What do you think of the latest changes?

There is one area that it is pretty difficult to get rid of the extra traversal. When using insert_all from a query: https://github.com/elixir-ecto/ecto/blob/fe2130587da04fbe1bf957d4e0c78e0511fd8516/lib/ecto/repo/schema.ex#L118

In order to remove the extra traversal you would have to detect inserts in `normalize_select` and then do the writable detection inside of `collect_fields`: https://github.com/elixir-ecto/ecto/blob/fe2130587da04fbe1bf957d4e0c78e0511fd8516/lib/ecto/query/planner.ex#L1392

I think that will get pretty messy.